src/Nethermind/Nethermind.TxPool/TxPool.cs

```diff
@@ -189,8 +189,51 @@ public AddTxResult AddTransaction(Transaction tx, TxHandlingOptions handlingOpti
             
             bool managedNonce = (handlingOptions & TxHandlingOptions.ManagedNonce) == TxHandlingOptions.ManagedNonce;
             bool isPersistentBroadcast = (handlingOptions & TxHandlingOptions.PersistentBroadcast) == TxHandlingOptions.PersistentBroadcast;
-            if (_logger.IsTrace) _logger.Trace($"Adding transaction {tx.ToString("  ")} - managed nonce: {managedNonce} | persistent brodcast {isPersistentBroadcast}");
+            if (_logger.IsTrace) _logger.Trace($"Adding transaction {tx.ToString("  ")} - managed nonce: {managedNonce} | persistent broadcast {isPersistentBroadcast}");
 
+            return FilterTransaction(tx, managedNonce) ?? AddCore(tx, isPersistentBroadcast);
+        }
+
+        private AddTxResult AddCore(Transaction tx, bool isPersistentBroadcast)
+        {
+            // !!! do not change it to |=
+            bool isKnown = _hashCache.Get(tx.Hash);
+
+            /*
+             * we need to make sure that the sender is resolved before adding to the distinct tx pool
+             * as the address is used in the distinct value calculation
+             */
+            if (!isKnown)
+            {
+                isKnown |= !_transactions.TryInsert(tx.Hash, tx);
+            }
+
+            if (!isKnown)
+            {
+                isKnown |= _txStorage.Get(tx.Hash) != null;
+            }
+
+            if (isKnown)
+            {
+                // If transaction is a bit older and already known then it may be stored in the persistent storage.
+                Metrics.PendingTransactionsKnown++;
+                if (_logger.IsTrace) _logger.Trace($"Skipped adding transaction {tx.ToString("  ")}, already known.");
+                return AddTxResult.AlreadyKnown;
+            }
+
+            tx.PoolIndex = _txIndex++;
+            _hashCache.Set(tx.Hash);
+
+            HandleOwnTransaction(tx, isPersistentBroadcast);
+
+            NotifySelectedPeers(tx);
+            FilterAndStoreTx(tx);
```

Comment:
why FilterAndStore is called after Notify?
is it the same filtering as the other filtering?

Reply:
Old code, not sure why it was named like this

Reply:
I know, I know, but maybe worth to name properly at the time when new filtering rules are introduced.